Exhaustively cover the `child_process` module

JordanMartinez opened the pull request:
**Description of the change**



---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

JordanMartinez committed:
Add AbortController code

JordanMartinez committed:
Exhaustively cover `child_process` module

JordanMartinez committed:
Reorder docs based on importance

JordanMartinez committed:
Fix mistake: send returns Boolean

JordanMartinez committed:
Use ImmutableBuffer; add missing option

JordanMartinez committed:
Add variants where options are ignored

JordanMartinez committed:
Fix return type of execFileSync

JordanMartinez committed:
Update tests

JordanMartinez committed:
Remove duplicate functions in FFI

JordanMartinez committed:
Fix linting error in abort controller FFI

JordanMartinez committed:
Add argonaut-core as dep

JordanMartinez commented:
I think the `AbortController` should be in a separate repo.

jamesdbrock commented:
> I think the `AbortController` should be in a separate repo.

There is an `AbortController` in https://pursuit.purescript.org/packages/purescript-web-fetch/3.0.0/docs/Web.Fetch.AbortController#t:AbortController

JordanMartinez commented:
> > I think the `AbortController` should be in a separate repo.
> 
> There is an `AbortController` in https://pursuit.purescript.org/packages/purescript-web-fetch/3.0.0/docs/Web.Fetch.AbortController#t:AbortController

Correct me if I'm wrong, but that's specific to a web context.

garyb commented:
Why use `Json` rather than `Foreign` in a few places in here? It may be that the value is basic enough to be represented as JSON, but it seems a little odd to me to vary between them.

JordanMartinez commented:
> > > I think the `AbortController` should be in a separate repo.
> > 
> > 
> > There is an `AbortController` in https://pursuit.purescript.org/packages/purescript-web-fetch/3.0.0/docs/Web.Fetch.AbortController#t:AbortController
> 
> Correct me if I'm wrong, but that's specific to a web context.

I'm wrong. The interfaces in a Node/Web context are the same:
- https://nodejs.org/api/globals.html#class-abortcontroller
- https://dom.spec.whatwg.org/#interface-abortcontroller

JordanMartinez committed:
Fix FFI: use AbortSignal for static methods

JordanMartinez commented:
> Why use `Json` rather than `Foreign` in a few places in here? It may be that the value is basic enough to be represented as JSON, but it seems a little odd to me to vary between them.

Because the docs said the following, I thought JSON was appropriate:
> message [Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) - A parsed JSON object or primitive value.
> https://nodejs.org/api/child_process.html#event-message

`Foreign` is arguably more consistent because that's what `send` is using... I'll make the change.

JordanMartinez committed:
Drop Json & use Foreign

JordanMartinez committed:
Add more type safety around 'serialization' option

JordanMartinez commented:
I think I'd like to track at the type-level whether `pipe` was provided to the `stdio` options, so that I can remove the `Maybe` wrapper around `stdin`/`stdout`/`stderr` and just make it callable or not. I can still expose a version that wraps the result in `Maybe` if one doesn't know what it will be but still needs to do something if it's there.

JordanMartinez committed:
Drop test npm script

JordanMartinez committed:
Revert "Drop test npm script"

This reverts commit ebadec4d40668563bee91cca222a0c87fdd12e23.

JordanMartinez commented:
Recent two commits were due to making the change on the wrong PR. I'm setting up the `js-abort-controller` repo.

JordanMartinez committed:
Add dep on js-abort-controller

JordanMartinez committed:
Drop unneeded Async name in js type alias

JordanMartinez committed:
Correct `onExit`: one value must be non-null

JordanMartinez committed:
Fix return type of spawnSync

JordanMartinez committed:
Review options: add/remove/fix incorrect ones

JordanMartinez committed:
Update tests to account for recent changes

JordanMartinez commented:
Ok, I've gone through all the options one more time now that I've had more sleep in me. I still need to do the `stdio` stuff, and I think I'll use some type-level stuff to achieve that goal. But, I'm out of time for now. Once that's done, this should be ready for a final review.

JordanMartinez committed:
Ensure stdin/stdout/stderr are type-safe

JordanMartinez committed:
Move send docs back to function

JordanMartinez commented:
I think I should add another type parameter that tracks whether the child process was started with `detached: true`. That can be used to determine whether `cp.unref()` or `cp.ref()` actually does anything, which AFAICT only does something when:
- the `detached: true` option was set, AND
- the `"ipc"` value was not set in the `stdio` option's array.

Once that's done, I think this is ready for a final review.

JordanMartinez committed:
Uncomment FFI exports now that PS compiles

JordanMartinez committed:
Prevent stream sharing with invalid streams

It's invalid to create a write stream
and then share that stream as
the child process' `stdin`.
Rather, the child process' `stdin`
must be a readable or duplex
stream.

JordanMartinez committed:
Update docs

JordanMartinez commented on src/Node/ChildProcess/StdIO.purs:
```diff
@@ -0,0 +1,206 @@
+-- |
+-- | The verbosity of the `ConfigureStdIoOptionsFn` type is needed
+-- | to make `stdin`/`stdout`/`stderr` type safe.
+toStdIoOption
+  :: forall stdInUsed stdOutUsed stdErrUsed ipcUsed
+   . ConfigureStdIoOptionsFn stdInUsed stdOutUsed stdErrUsed ipcUsed
+  -> StdIoOption stdInUsed stdOutUsed stdErrUsed ipcUsed
+toStdIoOption buildOptions = do
+  let
+    stdinOpt = maybe undefined stdIoToForeign $ map (\fn -> fn choicesStdin) options.stdin
+    stdoutOpt = maybe undefined stdIoToForeign $ map (\fn -> fn choicesStdOutErr) options.stdout
+    stderrOpt = maybe undefined stdIoToForeign $ map (\fn -> fn choicesStdOutErr) options.stderr
+    ipcOpt
+      | Just fn <- options.ipc
+      , IpcOption true <- fn noIpc = unsafeToForeign "ipc"
+      | otherwise = undefined
+
+    toStdIO :: Array Foreign -> StdIoOption stdInUsed stdOutUsed stdErrUsed ipcUsed
+    toStdIO = unsafeCoerce
+  toStdIO $ [ stdinOpt, stdoutOpt, stderrOpt, ipcOpt ] <> options.extra
+  where
+  options = buildOptions defaults
+
+  defaults =
+    { stdin: Nothing
+    , stdout: Nothing
+    , stderr: Nothing
+    , ipc: Nothing
+    , extra: []
+    }
+
+  stdIoToForeign :: forall r. StdIO r -> Foreign
+  stdIoToForeign = unsafeToForeign
```
A LOT is going on in this function, so let me clarify things for reviewers.

Goals:
1. We want to track at the type level whether a `stdin`/`stdout`/`stderr` stream (1) exists and (2) what can be done with it.
2. We also want these options to be easily overridden by the end-user. In other words, I must enjoy writing the options as an end-user.
3. We want to prevent hard-to-understand compiler errors or at least make it obvious what went wrong.
4. We also want to track whether the `ipc` stdio option was included in the array. This option can only appear at most once in the array, and it enables things like `send`. It's possible values are different than `stdin` and friends. That's why it's handled separately.
5. In the event that the user provides their own streams in `stdio` (hence the `extra` field), we want their access to such things to be consistent.

In pursuit of goal 2, I followed the "defaults" syntax. (e.g. `someFunction $ defaults { label = override }`. However, the `defaults` value is often already known and pointless. So, rather than requiring a value, we require a function that passes that default value into the function. This saves us a few characters and removes the need to import something:
```diff
-someFunction $ defaults { label = value }
+someFunction $ _ { label = value }
```

In pursuit of goal 1... The `stdin`/`stdout`/`stderr` are consumers. We want to get a `Stream x` from the child process (e.g. `Readable ()` for `stdin`). Thus, we need to track on the `ChildProcess` type what the stream type will be for `stdin`/`stdout`/`stderr`. If that value ends up being `Stream ()`, there's nothing anyone can do with that. So, it acts like a type-save `null`. Similarly, if it's `Stream (read :: Proxy "some message")`, one can't do anything with that either (more on why there's a `Proxy` there later). Thus, the value returned by this function stores what the `x` in `Stream x` must be for `stdin`/`stdout`/`stderr`.

If we use `inherit` for `stdio[0]`/`stdin`, the stream we get back is the process' `stdin`, which is a `Readable ()`. If we use `inherit` for `stdio[1]`/`stdout`, we get back the Process' `stdout`, which is a `Writeable ()`. To overload both possibilities and only need to use one value `inherit`, so that it matches Node's docs but is still type safe, we use a callback function.
```
-- type-annotated value for 'inherit'
inheritRead :: StdIO (read :: Read)
inheritRead = unsafeCoerce "inherit"

inheritWrite :: StdIO (write :: Write)
inheritWrite = unsafeCoerce "inherit"

-- select the choice from a record of options
inherit :: forall r. { useInherit :: StdIO x | r } -> StdIO x
inherit { useInherit } = useInherit

type StdInChoices =
  { ... other choices
  , useInherit :: StdIO (read :: Read)
  }
  
type StdOutErrChoices =
  { ... other choices
  , useInherit :: StdIO (write :: Write)
  }
```
For `stdin`, we pass in a record of `StdInChoices`, which ensures that `inherit` will select the `(read :: Read)` row. For `stdout`, we pass in the `StdOutErrChoices`, where `inherit` will select the `(write :: Write)` row. The reason why the row polymorphism is needed here is because some streams that work in `stdin` context (e.g. read-only or read-write streams) don't work in `stdout` context (e.g write-only or read-write streams).

For goal 3, we use monomorphic records for `StdInChoices`/`StdOutErrChoices` to prevent the need to use `RowList` (which has difficulties with polymorphic record inference). This prevents some possible errors.

Lastly, for goal 3, `ignore` will provide a row of `(read :: Proxy "some msg", write :: Proxy "some msg")`. In the event that someone still tries to use `write (stdin childProcess) ...` when they set `stdin` to `ignore`, the message they will get will be a type unification between
```
read :: Read
vs
read :: Proxy "'ignore' was used for this IO slot, so this stream on the child process doesn't exist."
```
which should make it obvious what went wrong. Feedback on what would be a good error message would be helpful as well as where else I should replace a row of `()` with `(read :: Proxy msg, write :: Proxy msg)`. The reason both rows are supplied is to ensure the message appears regardless of which stream function one calls (e.g. `read/write`).
However, upon writing this, I realize that `Stream.onClose` doesn't verify whether the rows in the Stream are `read` or `write`. Hmm... :thinking:...

For goal 4, we just newtype a Boolean and use `"ipc"` or `undefined` depending on that boolean. Per my reading of the docs, `fork` requires one to include `"ipc"` somewhere in the stdio array, so that function overrides the defaults with `true` and forces the user to return `true` for this option.

For goal 5, the array's first 4 values will always correspond to `[ stdin, stdout, stderr, ipc ]` and each value is either undefined or one of the options (e.g. `inherit`, `pipe`, etc.) described in Node's docs. Thus, anything one wants to add after that will be at the 5th index (i.e. `stdio[4]`) or higher. If one wants to further track things, they will probably newtype `ChildProces` so they can access the corresponding slot in the `stdio` array with type-safety guarantees.

JordanMartinez replied:
> ignore will provide a row of (read :: Proxy "some msg", write :: Proxy "some msg")....
> ...
> However, upon writing this, I realize that `Stream.onClose` doesn't verify whether the rows in the Stream are read or write. Hmm... :thinking: ....

The issue I was raising above is that my previous approach would allow one to use `ignore` and then call `Stream.close (stdout cp)` even though `stdout` in that usage would be `null`. The latest commit updates `ChildProcess` and `StdIO` to store the full type. So, using `ignore` means calling `stdout cp` now returns `Void`.

JordanMartinez committed:
Fix bug: 'ignore' can close things not streams

Since `ignore` would return ` Stream r`, one could call `Stream.close` on it.
By just storing the value that `stdin`, etc.,
will be when one access it, we can forego
this problem altogether

JordanMartinez commented:
> I think I should add another type parameter that tracks whether the child process was started with `detached: true`. That can be used to determine whether `cp.unref()` or `cp.ref()` actually does anything, which AFAICT only does something when:
> 
>     * the `detached: true` option was set, AND
> 
>     * the `"ipc"` value was not set in the `stdio` option's array.
> 

This is a lot harder than just that. On Windows, setting `detached: true` means the child process will continue running even after the parent exists. On non-Windows, the child process can continue running even after the parent exits whether `detached` is set to `true` or not.

I don't think I care enough about this option to figure out whether it can be done, nor how.

JordanMartinez committed:
More docs tweaks

JordanMartinez commented:
This is ready for a review.

JordanMartinez commented:
Closing now that I've provided my own bindings via `purescript-node-execa`.
